packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts

```diff
@@ -482,8 +514,25 @@ export async function visualizeChildData(
  * Determines whether or not the provided value is an {@link ISharedObject}, for the purposes of this library.
  * @remarks Implemented by checking for the particular properties / methods we use in this module.
  */
-function isSharedObject(value: Partial<ISharedObject>): value is ISharedObject {
+function isSharedObject(value: unknown): value is ISharedObject {
+	return (
+		(value as SharedObject).id !== undefined &&
+		(value as SharedObject).attributes?.type !== undefined &&
+		(value as SharedObject).on !== undefined
+	);
+}
+
+/**
+ * Determines whether or not the provided value is an {@link DataObject}, for the purposes of this library.
+ * @remarks Implemented by checking for the particular properties / methods we use in this module
+ */
+function isDataObject(value: unknown): value is DataObject {
 	return (
-		value.id !== undefined && value.attributes?.type !== undefined && value.on !== undefined
+		(value as DataObject).initializeInternal !== undefined &&
+		(value as DataObject).id !== undefined &&
+		(value as DataObject).IFluidLoadable !== undefined &&
+		(value as DataObject).IFluidHandle !== undefined &&
+		(value as DataObject).handle !== undefined &&
+		(value as VisualDataObject).root.attributes.type === DirectoryFactory.Type
```

Comment:
Since we're not really sure if `root` even exists, we need to be defensive when accessing sub-properties, or this could throw when trying to access a property of an `undefined`.
```suggestion
		(value as VisualDataObject).root?.attributes?.type === DirectoryFactory.Type
```

Reply:
Interesting, the reason why I excluded undefined check is:
-  `DataObject.root` *can* be *empty* but should not be `undefined` when accessing the value (check [here](https://github.com/microsoft/FluidFramework/blob/8c926a73bde358d3f2f513ae184b799737fc66a4/packages/framework/aqueduct/src/data-objects/dataObject.ts#L38))
- Thus, if the `value` typecasted to `VisualDataObject` (i.e., `DataObject`) does not have `root` defined, this typeguard function should fail. 

Am I missing something?

Reply:
> if the value typecasted to VisualDataObject (i.e., DataObject) does not have root defined, this typeguard function should fail.

I think it should "fail" by returning `false` instead of throwing an error trying to access a property of an `undefined` though. We would know for sure that the input is not a data object, so the function just saying that instead of throwing would be ok.

Also, from the perspective of a typeguard function, I think the point is that we have no idea of what the thing we received actually is or what shape it has (thus why it receives as `unknown`). Whether the `DataObject` class validates that the thing it returns as `root` is there or not doesn't matter for the logic in the typeguard; maybe this function gets called with a string input, and we don't want it to throw, just to return `false` in that case.